eth2util/deposit/deposit_data.go

```diff
@@ -0,0 +1,155 @@
+// Copyright © 2022 Obol Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package deposit
+
+import (
+	"encoding/hex"
+	"encoding/json"
+
+	eth2p0 "github.com/attestantio/go-eth2-client/spec/phase0"
+	ssz "github.com/ferranbt/fastssz"
+
+	"github.com/obolnetwork/charon/app/errors"
+)
+
+const depositAmt = 32000000000
+
+// DepositData contains all the information required for activating validators on the Ethereum Network.
+type DepositData struct { //nolint:revive
+	// PubKey is the group public key for a Distributed Validator.
+	PubKey eth2p0.BLSPubKey
+
+	// Amount is the amount of Eth needed to activate a validator.
+	Amount eth2p0.Gwei
+
+	// Eth1WithdrawalAddress is the Ethereum withdrawal address.
+	Eth1WithdrawalAddress string
+
+	// DepositMessageRoot is the hash tree root of DepositMessage.
+	DepositMessageRoot eth2p0.Root
+
+	// Signature is constructed from DepositMessageRoot combined with DOMAIN_DEPOSIT.
+	Signature eth2p0.BLSSignature
+
+	// ForkVersion identifies the network/chainID.
+	ForkVersion eth2p0.Version
+}
+
+func (d DepositData) HashTreeRoot() ([32]byte, error) {
+	b, err := ssz.HashWithDefaultHasher(d)
+	if err != nil {
+		return [32]byte{}, errors.Wrap(err, "hash deposit data")
+	}
+
+	return b, nil
+}
+
+func (d DepositData) HashTreeRootWith(hh *ssz.Hasher) error {
+	idx := hh.Index()
+
+	// Field 0 'PubKey`
+	hh.PutBytes(d.PubKey[:])
+
+	// Field 1 'Amount'
+	hh.PutUint64(uint64(d.Amount))
+
+	// Field 2 'Eth1WithdrawalAddress'
+	hh.PutBytes([]byte(d.Eth1WithdrawalAddress))
```

Comment:
this should probably be credentials, not address

Reply:
correct, also changed the ordering of fields according to spec: https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#depositdata